Zephyr ai article - Fixing link in Rust AI Article (#730)
espressif/developer-portal: content/blog/2026/06/zephyr-coding-with-ai/index.md

---

> title: "Developing a Zephyr IOT app with AI"

Reviewer: perhaps `IoT` instead.

Author: Fixed.

date: 2026-06-15
tags:
  - how-to
  - Zephyr
  - IoT
  - AI
  - LLM
showTableOfContents: true
authors:
  - ricardo-tafas

> summary: "Still on improving our AI-based, embedded software workflow, this article applies the same AI-agent discipline to Zephyr. Using the same M5 stack ESP Dualkey board, specs, integration rules,plan→execute→commit→test, reusable modules, journals, and Git - without redefining the product."

Reviewer: missing space: `rules, plan→execute→commit→test`

Reviewer: perhaps you could remove this non-ascii `→` symbol

Author: it rendered well. I'll keep it.

> summary: "Still on improving our AI-based, embedded software workflow, this article applies the same AI-agent discipline to Zephyr. Using the same M5 stack ESP Dualkey board, specs, integration rules,plan→execute→commit→test, reusable modules, journals, and Git - without redefining the product."

Reviewer: There are many places in the text where the dualkey board is referenced. Everytime it is written in different ways: M5 stack ESP Dualkey, M5Stack ESP-DualKey, ESP DualKey, ....

suggestion: keep always `M5Stack ESP-DualKey`

Reviewer 2: I also noticed discrepancies.

Actually, I just googled "ESP dualkey" and found that only the Espressif website and documentation have [ESP-DualKey](https://docs.espressif.com/projects/esp-dev-kits/en/latest/esp32s3/esp-dualkey/user_guide.html). The M5Stack only has [Chain DualKey](https://docs.m5stack.com/en/chain/Chain_DualKey).

@rftafas To avoid confusion with what product you used, can you please fix its spelling throughout the article and provide a link to the product page?

Author: This is tricky. On the box I have says ESP DualKey. Chain DualKey is because it is part of the "Chain" family. So, I'll settle for "ESP Dualkey". I also made it consistent on the prior kit. When M5 comes in the game, I"ll be using ESP DualKey from M5 Stack.

> summary: "Still on improving our AI-based, embedded software workflow, this article applies the same AI-agent discipline to Zephyr. Using the same ESP Dualkey kit from M5 Stack, specs, integration rules, plan→execute→commit→test, reusable modules, journals, and Git - without redefining the product."

Reviewer 3: Change from `M5 Stack` to `M5Stack`. remove the space.
There are more places where you have `M5 Stack`.

---

## Introduction

In [a previous article](https://developer.espressif.com/blog/2026/04/developing-a-rust-iot-app-with-ai/) I took the M5Stack ESP-DualKey through a journey of Rust firmware development. The goal? Porting the Espressif's Unified Provisioning over BLE and SoftAP from ESP-IDF and building a thing for the internet. Even though the project went as expected and I got the thing; the workflow mattered much more.

This post is about that workflow, now, applied to Zephyr. What I plan to elaborate here is not about becoming a Zephyr expert. Even though I do have some experience by being the Product Manager - and occasional tech support - of Zephyr inside Espressif, the goal is not Zephyr. It is how to work differently, with AI, beyond the fuss. The durable payoff is how to work with an agent - specs, boundaries, Git, evidence - and not which language or RTOS wins on paper. **Code is fast now** - maybe too fast - and the hard part becomes managing code and everything around it: architecture, organization, testing, and knowing when to stop the model from typing. Writing code goes away, but thinking in terms of "code" is still there.

> This time, it will be the same bet on the same board: DualKey still had to talk to the **Espressif ESP BLE Provisioning** app the same way before. I just moved the implementation to Zephyr C and got strict about how we work with the agent - clearer specs, sharper boundaries, more discipline around Git and the bench. If you have not read the Rust article yet, [start there](https://developer.espressif.com/blog/2026/04/developing-a-rust-iot-app-with-ai/) and come back; the product story will make more sense. 

Reviewer 2: ```suggestion
This time, it will be the same bet on the same board: ESP-DualKey still has to talk to the **Espressif ESP BLE Provisioning** app the same way as before. I just moved the implementation to Zephyr C and got strict about how we work with the agent - clearer specs, sharper boundaries, more discipline around Git and the bench. If you have not read the Rust article yet, [start there](https://developer.espressif.com/blog/2026/04/developing-a-rust-iot-app-with-ai/) and come back; the product story will make more sense. 
```

Author: Applied.

Author: agreed.

Finally, and more importantly, applying the lessons learned. No dumping of journals this time.

> ---

Reviewer 2: A line separating chapters reminds very much of AI responses.

Here and in the rest of the article, I suggest that you don't use `---` to avoid unnecessary associations with AI content generation.

```suggestion
```

Author: I indeed as AI to do some beautification. But the lines were my won. I'll remove them.

This experiment did not start from zero. On the agent side it was the same **Cursor** habit as before - Plan mode when the work was too big. Coding styles, adding project boundaries, structure, and so on.

We still had the **product spec** from the Rust work and the **Rust tree** as a behavioral reference - what the buttons should do, how provisioning should feel, what “done” meant on the bench. We re-used as much as we could from ESP-IDF and Rust.

What had to be **new** was the presence of a new OS: Zephyr. Conventional Zephyr project activities, tasks, methods, processes. All with AI. Thus, speaking about Zephyr...

### About Zephyr

**Zephyr** is an open-source real-time operating system from the [Zephyr Project](https://www.zephyrproject.org/) (Linux Foundation): one kernel and a large in-tree catalogue of drivers, protocol stacks, and services, configured with **Kconfig** and **devicetree** and built through a **west** metatool. It ties the RTOS, board support, libraries and your applications together. Boards and SoCs are integrated upstream; product firmware and reusable modules usually live in separate repositories rather than inside the OS tree. 

> **Zephyr on Espressif silicon:** Espressif offers and supports [Zephyr for ESP32-family SoCs](https://docs.espressif.com/projects/esp-zephyr/en/latest/) - the same M5Stack ESP-DualKey hardware as the Rust firmware we built before. That made this a platform shift on familiar silicon, not a new board story. I mention this once here and do not repeat the portal pitch later.

Reviewer 2: The link `https://docs.espressif.com/projects/esp-zephyr/en/latest/` returns 404.

Author: Fixed.

> **Zephyr on Espressif silicon:** Espressif offers and supports [Zephyr for ESP32-family SoCs](https://www.espressif.com/en/sdks/esp-zephyr). In Espressif, we consider Zephyr to be an OS Distribution that can run on Espressif computers. It is consdiered a solid solution and since 2020 we have been working on it. The current status of Zephyr can be found in the [Status Page](https://developer.espressif.com/software/zephyr-support-status/), on this portal.

Reviewer: typo: `considered`

Author: fixed.

> In the case of Zephyr, it was easier to have it previously installed and configured. Zephyr installation is not a big problem and it is fairly well documented. I believe AI could do it, but this is a test for later.

Reviewer 2: ```suggestion
In the case of Zephyr, I already had it installed and configured. Zephyr installation is not a big problem and it is fairly well documented. I believe AI could do it, but this is a test for later.
```

Author: agreed.

### Why Zephyr this time

As stated, Espressif is already supporting Zephyr. Zephyr adoption among Espressif users has been growing steadily. More and more people are considering the Espressif-Zephyr duo for their projects. Naturally, at Espressif, we aim to give users a great experience and good resources. However, we have seen customers complain about the difficulties of starting a new project from scratch. 

> This was smelling like a good challenge for a new approach. We have references, we did it for another design, the phone app and reference C define what “correct” means; Zephyr code implements, it does not redefine, the protocol. Because of all this, we chose Zephyr to keep evolving the process; not our knowledge of firmware and not because Zephyr replaces Rust everywhere. They are great solutions. As ESP-IDF is. 

Reviewer 2: Instead of `was smelling`, I recommend using something more idiomatic:

```suggestion
This felt like a good challenge for a new approach. We have references, we did it for another design, the phone app and reference C define what “correct” means; Zephyr code implements, it does not redefine, the protocol. Because of all this, we chose Zephyr to keep evolving the process; not our knowledge of firmware and not because Zephyr replaces Rust everywhere. They are great solutions. As ESP-IDF is. 
```

Author: Agreed.

### Product definition as the goal

We carried [`docs/spec/product_spec.md`](https://github.com/rftafas/dualkey-provisioning/blob/main/docs/spec/product_spec.md) forward and adjusted it where Zephyr mattered - gestures, LEDs, provisioning entry and exit (including reboot policy), MQTT and HID, explicit out-of-scope lines. It was not a new product definition; it was the same contract on a new stack.

With AI, that document is the **goal** (as in the [earlier article on specs](https://developer.espressif.com/blog/2026/04/developing-a-rust-iot-app-with-ai/)): SSID and BLE names, timeouts, cancel at product level, what “done” means on the bench. When spec and code disagree, you fix one or the other deliberately - never both drifting.

---

> ## The contracts (or design rules)

Reviewer 2: The subsections here:

1. What we already knew, what the Rust article wrote down
2. Habits we already had
3. What the Rust article added (mostly naming and packaging)
4. Zephyr integration principles (what Zephyr added to the contract)

Questions:

- The headings 1 and 3 are overlapping. Is heading 1 really needed?
- Shouldn't the logic be: (1) Baseline habits/knowledge, (2) What Rust project brought about, (3) What Zephyr project brought about?

Author: Leftover after my reorganization of article paragraphs. First heading removed as unnecessary.

1. **Use as much Zephyr as possible** - prefer the RTOS and its supported flows over parallel “mini-OS” layers in the app.
2. **Use as much of the OS services as possible** - networking, Wi‑Fi management, BLE host, settings/NVS patterns, logging, workqueues: wire through Zephyr subsystems instead of bespoke shims where a subsystem already exists.
3. **Use as much of the Zephyr distro as possible** - leverage in-tree modules and Kconfig-selected building blocks before private copies; contribute upstream when something is missing rather than forking silently.
4. **Integrate as seamlessly to Zephyr as possible** - west workspace, out-of-tree module layout, devicetree, `prj.conf`, samples, naming and docs that look like a normal Zephyr integration.

Those four lines are **excellent prompt constraints**: they tell the model what “good” looks like before it invents a foreign project shape. Later we show how they showed up in repositories and build artifacts without re-deriving the list.

**Three layers:** product behavior (`product_spec.md`), Zephyr integration principles (this list), then the new repos and build layout (later).

---

> ## New lessons for the contract (this experiment)

Reviewer 2: The relation between these is not exactly clear to me:

- Zephyr integration principles (what Zephyr added to the contract) -- where these additions to the contract come from? Did you deduce them from your experience?
- New lessons for the contract (this experiment) -- do these amendments come from your retrospective analisys?

I would recommend that you make it more clear in the headings and introductions at least. Otherwise, the direction in which the narrative goes is not clear.

Author: Agreed. I smoothed our the text. To address the question, before first article were guesses and prior knowledge. Between the 2 articles, I had the article 1 experience added on top. And now I have even more added on top. There is space for a third.

> If the contract above is the constitution, things we brought from prior Firmware experience and teachings from the Rust experiment, what follows next are the amendments after Zephyr provisioning experiment. Many more aspects were revealed thorough this activity, things that were new or things that had their meaning modified. Experience, as they say.

Reviewer: suggestd: `revealed thorough this`

| Suggesting API shapes | Whether a public API is stable and minimal |
| Reading source quickly | What the trace proves happened on the chip |
| Large diffs | Whether a large diff is allowed this milestone |

This experiment drove that home: when provisioning misbehaved, the winning move was still “paste the log, state the expectation, ask what branch is wrong” - not “generate more code until it compiles.”

### Organization is the human’s job

Entropy from the Rust article is still true: throwaway scripts, duplicate docs, and “helpful” files in the wrong folder appear unless someone enforces structure. **You** own spec versus journal, module versus app, `.gitignore`, submodule pins, and what gets deleted versus committed.

**Dedupe** is part of that job: when a module manual exists, stop re-exporting the same prose in the product repo. We kept that discipline this time (contrast the lost journal from the Rust cleanup).

### Stop letting AI own the whole tree - extract a component

There was an “AI does everything in one tree” phase. This time we pivoted to a **reusable component**, a **proof application**, and a **thin product** - stack-agnostic, and explicit about reusability.

> **Macaronic code (generic fix):** We had already named the pitfall in the Rust write-up - logic piled in one tree, weak boundaries, structures the model will not invent unless you force them. Our answer this time is architectural, not syntactic: a **formal component** and a **formal app**, a normative component spec separate from a debug journal, and a public API reviewed like a library. The esp-provisioning module and DualKey repos are the Zephyr-shaped instance; the lesson is not Zephyr-specific.

Reviewer 2: ```suggestion
**Macaronic code (generic fix):** We had already named the pitfall in the Rust write-up - logic piled in one tree, weak boundaries, structures that the model will not invent unless you force them. Our answer this time is architectural, not syntactic: a **formal component** and a **formal app**, a normative component spec separate from a debug journal, and a public API reviewed like a library. The esp-provisioning module and DualKey repos are the Zephyr-shaped instance; the lesson is not Zephyr-specific.
```

Author: Agreed.

**AI does not handle upstream Git.** The agent can prepare diffs and summaries; a human lands upstream pull requests. Sometimes, the trains and airplanes need human supervision. Things can go wildly wrong.

**Journals yes, a hard yes.** We **kept** `journal.md` as a localized "troubleshooting and regression avoider". I also found that it was preventing re-reasoning. AI is a statistical machine; after a while, it might try the same trick that didn't work before. The journal prevented it. It worked like a charm in far more ways than expected. They are in the module repo this time. First thing a plan needs to include is _to update the journal_.

Usually, that is exactly how I do it: during planning, I request: _todo: update the journal with all positive results from previous loop and the solutions that were tried and didn't solve the problem_. I like to add the problem and what we did to fix it, in other words, the problem and the fix go to the journal. Many times AI undoes work, goes to Journal, finds an entry - directly or similar - and in the same round, fixes it.

---

> ## Plan → execute → commit → test (master gear)

Reviewer 2: This section looks like an odd one out. How is it related to the previous and the following ones?

Maybe you can make the transitions more clear.

Author: This is the core of the process. It is a lesson so important that I'm creating a whole session for it.

In terms of 'location', it is just as a lesson as the others. But for me it is becoming base of an AI-based dev process. Master gear, as I said.

This rhythm is **generic** - it applies to Rust, Zephyr, or anything else. Zephyr shows up only as examples elsewhere.

The Rust article mentioned planning for long sequences but did not operationalize what happens **after** a plan runs. We elevated that to its own gear:

1. **Plan** before non-trivial agent work (Cursor Plan mode or an explicit plan document).
2. **Execute** the plan - agent, human, or both.
3. **Commit** a reversible checkpoint *before* the messy part of testing when the change is still easy to roll back - because on the bench the agent (or you) can go wild trying “one more fix.”
4. **Test** on the bench: build, flash, smoke scenario; for components, exercise the **sample harness** before the full product.
5. **Loop** after testing, **restart planning** with the results of the prior phase rather than stacking unbounded “fix it” prompts on a failed run. Seriously, go back to planning. Yes, the temptation to let AI start coding wildly and fix the bug is there. It should be resisted. This is when AI-produced code becomes difficult to support and maintain.

> As far as process goes, this is PDCA - plan, do, check, act - wearing firmware clothes. From first article, incredibly like managing Junior Developers. With the whole development process in git, it will be easy to review the architecture. Furthermore, we can always squash commits and review the messages - the plans - of what really was kept between commits. It is a game changer in traceability.

Reviewer 2: > From first article, incredibly like managing Junior Developers.

Please clarify this sentence.

Author: Clarified. Actually, I added some smooth transition of ideas.

4. **Test** on the bench: build, flash, smoke scenario; for components, exercise the **sample harness** before the full product.
5. **Loop** after testing, **restart planning** with the results of the prior phase rather than stacking unbounded “fix it” prompts on a failed run. Seriously, go back to planning. Yes, the temptation to let AI start coding wildly and fix the bug is there. It should be resisted. This is when AI-produced code becomes difficult to support and maintain.

> As far as process goes, this is PDCA - plan, do, check, act - wearing firmware clothes. That holds up. From first article, I commented that working with AI was incredibly similar to managing Junior Developers on a technical level, a feeling that reinforces this PDCA concept. By adding the whole development process in git with frequent commits, it will be easy to review the solution, or to control it. 

Reviewer 2: We created a new CI check to notify about trailing spaces. Some authors have the habit of using two spaces to create a line break in markdown. Many engineers have IDE rules to remove trailing spaces, multiple new lines from the end of the file, etc. The two approaches don't work together very well. Where new lines are needed, we encourage using `<br>`. Otherwise, there should be no trailing spaces, even just one like in your case.

Please remove the trailing space.

```suggestion
As far as process goes, this is PDCA - plan, do, check, act - wearing firmware clothes. That holds up. From first article, I commented that working with AI was incredibly similar to managing Junior Developers on a technical level, a feeling that reinforces this PDCA concept. By adding the whole development process in git with frequent commits, it will be easy to review the solution, or to control it.
```

> As far as process goes, this is PDCA - plan, do, check, act - wearing firmware clothes. That holds up. From first article, I commented that working with AI was incredibly similar to managing Junior Developers on a technical level, a feeling that reinforces this PDCA concept. By adding the whole development process in Git with frequent commits, it will be easy to review the solution, or to control it.

Reviewer 3: PDCA - plan, do, check, act -> PDCA - Plan, Do, Check, Act

> Git stays the safety net; commit granularity and artifact habits keep bisect honest; this loop is the **operating rhythm** that makes the contract stick—plan is not a chat luxury, and each execution ends in verify and commit. Before any comments that this will result in a polluted GIT history, I recall that we can always squash commits and review the messages - the plans - of what really was kept between commits. It is a game changer in traceability and an increase in entropy, but a controlled one.

Reviewer: suggestion: `Git` instead of `GIT`

Again, we learned a few tricks as well. And again, we found some Zephyr related teachings that I believe can be made generic and applicable to any software we use as our base.

### Zephyr integration patterns we used

Here it goes:

- **Module layout:** follow best practices. Not new, but always worth mentioning.
- **Try the sample apps.** Life saving, time saving decision. Lots of energy is saved when trying every component individually. 
- **Product app.** Have it organized. Business logic from basic component support.
- **Don't touch Upstream Code.** AI had the very bad habit to fix the issues or add tracing to the first piece of software it found convenient. This includes a lot of edits to main sources of Zephyr. It cannot differentiate code, code is code. I kept a hard rule here.
- **Except when you have to touch upstream content.** That is the beauty. It asked my permission to edit code because it was confident that new Zephyr code was necessary. This is when I went mostly old school and really reviewed the code. When you might interfere with a whole project, yes, you - at least I - become more careful.

### DualKey / esp-provisioning outcome (brief)

Public repositories:

> - [zephyr/modules/esp-provisioning](https://github.com/rftafas/dualkey-provisioning/tree/main/zephyr/modules/esp-provisioning) - out-of-tree module and shell sample.

Reviewer 2: The link returns 404.

Author: fixed.

- [zephyr/modules/esp-provisioning](https://github.com/rftafas/esp-provisioning-zephyr) - out-of-tree module and shell sample.
- [zephyr/](https://github.com/rftafas/dualkey-provisioning/tree/main/zephyr) - product firmware consuming the module.

On the bench, the shell sample exercises BLE and SoftAP transports with the Espressif provisioning app; the integrated app follows the same product spec. The module API was trimmed over time to stay library-shaped.

## Conclusion

The pair-programming split from the Rust article still holds: the model codes; I evaluate - against spec, reference trees, build output, UART traces, and the bench. **Code is cheap, judgment is not.** What changed as we kept writing is that we are slowly **consolidating a process** around AI-driven development instead of treating each chat as a one-off. Specs, rules, journals, reusable modules, proof apps before products, commit granularity, and **plan → execute → commit → test** are not trendy extras - they are how you get a real **productivity boost** without paying for it in regressions nobody can bisect.

> The lessons stack. The Rust pass gave us the duo contract, entropy control, and the warning about macaronic trees and lost notebooks. The Zephyr pass added platform-shaped components, sharper API boundaries, proof-before-product on the bench, and a human-owned path for upstream Git. None of that replaces the old discipline we already knew. **PDCA** - plan, do, check, act - is still there, just wearing Cursor and a faster typist. Git as undo, bounded tasks, UART as ground truth - still there too. The challenge is not to invent a new religion every time the stack changes. It is to **reuse what we already know**: carry the product spec forward, point the agent at prior source and reference code, write the rules down once, and keep the artifacts the next session can read. Start from scratch and you repeat the same errors - deleted journals, bloated trees, “it should work” without a trace.

Reviewer 3: Same here: PDCA - plan, do, check, act -> PDCA - Plan, Do, Check, Act

Author: redundant. Kept just the first one.

So the work ahead is curation as much as coding: tighten the contract, train the agent with project standards, and let it accelerate the keyboard while you keep the evaluator seat, the strategist and the Captain Navigator, that points where the ship should go. That is where the speed actually lands.

## Links

- [Developing a Rust-based IoT device with AI](https://developer.espressif.com/blog/2026/04/developing-a-rust-iot-app-with-ai/) (April 2026)
- [dualkey-provisioning — rust/](https://github.com/rftafas/dualkey-provisioning/tree/main/rust) (Rust firmware)

> - [zephyr/modules/esp-provisioning](https://github.com/rftafas/dualkey-provisioning/tree/main/zephyr/modules/esp-provisioning)

Reviewer 2: 404

Author: fixed.
